segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/metadata/TableBasedMetadataStore.java

```diff
@@ -196,6 +203,56 @@ public TableBasedMetadataStore(String tableName, TableStore tableStore, ChunkedS
                 });
     }
 
+    /**
+     * Retrieve all entries.
+     *
+     * @return A CompletableFuture that, when completed, will contain {@link Stream} of {@link StorageMetadata} entries.
+     * If the operation failed, it will be completed with the appropriate exception.
+     */
+    @Override
+    public CompletableFuture<Stream<StorageMetadata>> getAllEntries() {
+        return this.tableStore.entryIterator(tableName, IteratorArgs.builder().fetchTimeout(timeout).build())
+                .thenApplyAsync( i -> getStreamFromTableIterator(i).map(td -> td.getValue()), getExecutor())
+                .exceptionally(e -> {
+                    val ex = Exceptions.unwrap(e);
+                    throw new CompletionException(handleException(ex));
+                });
+    }
+
+    /**
+     * Retrieve all keys.
+     *
+     * @return A CompletableFuture that, when completed, will contain {@link Stream} of  {@link String} keys.
+     * If the operation failed, it will be completed with the appropriate exception.
+     */
+    @Override
+    public CompletableFuture<Stream<String>> getAllKeys() {
+        return this.tableStore.entryIterator(tableName, IteratorArgs.builder().fetchTimeout(timeout).build())
+                .thenApplyAsync( i -> getStreamFromTableIterator(i).map(td -> td.getKey()), getExecutor())
+                .exceptionally(e -> {
+                    val ex = Exceptions.unwrap(e);
+                    throw new CompletionException(handleException(ex));
+                });
+    }
+
+    /**
+     * Converts the given {@link AsyncIterator} to {@link Stream} of {@link TransactionData}.
+     * @param iterator Iterator to convert.
+     * @return {@link Stream} of {@link TransactionData}.
+     */
+    Stream<TransactionData> getStreamFromTableIterator(AsyncIterator<IteratorItem<TableEntry>> iterator) {
+        return StreamSupport.stream(Spliterators.spliteratorUnknownSize(iterator.asIterator(), Spliterator.ORDERED), true)
```

Comment:
What is the expected ordering of this Stream (if any)? Using the "parallel" flag may imply that ordering is going to be hard to guarantee upon consumption. Is ordering actually needed for the Table Entries read as a result?

Reply:
If order is not guaranteed, just note that in the description of the method please.

Reply:
Technically we don't really care about ordering. 
I'll change that.

Reply:
fixed.